[orchagent] add & remove port counters dynamically each time port was added or removed

tomer-israel opened the pull request:
**What I did**
add support for adding & removing port counters dynamically each time port was added or removed
the counters that were supported are:
- pg watermark counters
- pg drop counters
- queue stat counters
- queue watermark counters
- debug counters

* buffer drop counters and port stat counters are already supported to be added or removed each time port  is added/removed

**Why I did it**
in order to support dynamically add or removed ports on sonic

**How I verified it**
tested manually that the flex counters were added or removed correctly whenever we add or remove ports
added new test cases to the following vs tests:
- test_flex_counters.py
- test_drop_counters.py
- test_pg_drop_counter.py

**Details if related**

tomer-israel committed:
countrs support for dynamic port add/remove

tomer-israel committed:
countrs support for dynamic port add/remove

tomer-israel committed:
countrs support for dynamic port add/remove

tomer-israel committed:
countrs support for dynamic port add/remove

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -2411,14 +2423,34 @@ void PortsOrch::deInitPort(string alias, sai_object_id_t port_id)
     Port p(alias, Port::PHY);
-    p.m_port_id = port_id;
+
+    if (!getPort(port_id, p))
```
will be changed

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
 
+void PortsOrch::removeQueueMapPerPort(const Port& port)
+{
+    /* Create the Queue map in the Counter DB */
+    /* Add stat counters to flex_counter */
```
will be removed

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
+            queueIndexVector.emplace_back(id);
+        }
+
+        // Install a flex counter for this queue to track stats
```
will be removed

tomer-israel replied:
removed

tomer-israel commented on orchagent/debugcounterorch.cpp:
```diff
@@ -616,3 +637,50 @@ bool DebugCounterOrch::isDropReasonValid(const string& drop_reason) const
+
+    SWSS_LOG_NOTICE("add debug counter for port 0x%" PRIu64 , port_id);
+
+    for (auto it = debug_counters.begin(); it != debug_counters.end(); it++) {
```
changed

tomer-israel commented on orchagent/debugcounterorch.cpp:
```diff
@@ -616,3 +637,50 @@ bool DebugCounterOrch::isDropReasonValid(const string& drop_reason) const
+    SWSS_LOG_NOTICE("add debug counter for port 0x%" PRIu64 , port_id);
+
+    for (auto it = debug_counters.begin(); it != debug_counters.end(); it++) {
+        DebugCounter *counter = dynamic_cast<DebugCounter*>(it->second.get());
```
changed

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
+    vector<string> queueIndexVector;
+    vector<string> queueTypeVector;
+
+    for (size_t queueIndex = 0; queueIndex < port.m_queue_ids.size(); ++queueIndex)
```
i dont understand.
this loop was taken from the add command (not new code)

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
+        }
+        queue_stat_manager.clearCounterIdList(port.m_queue_ids[queueIndex]);
+
+        /* add watermark queue counters */
```
this comment is relevant for the generatePriorityGroupMapPerPort function in the removePriorityGroupMapPerPort i will change it to remove watermark queue...

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
+    m_counter_db->hdel(COUNTERS_QUEUE_INDEX_MAP, queueIndexVector);
+    m_counter_db->hdel(COUNTERS_QUEUE_TYPE_MAP, queueTypeVector);
+
+    CounterCheckOrch::getInstance().removePort(port);
```
it's possible, i just wanted to follow the current implementation. in the generatePriorityGroupMapPerPort function we are using:
CounterCheckOrch::getInstance().addPort(port);
so in the remove function we will use:
CounterCheckOrch::getInstance().removePort(port);

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5340,6 +5371,55 @@ void PortsOrch::generateQueueMap()
+        m_flexCounterTable->del(key);
+    }
+
+    m_counter_db->hdel(COUNTERS_QUEUE_NAME_MAP, queueVector);
+    m_counter_db->hdel(COUNTERS_QUEUE_PORT_MAP, queuePortVector);
+    m_counter_db->hdel(COUNTERS_QUEUE_INDEX_MAP, queueIndexVector);
+    m_counter_db->hdel(COUNTERS_QUEUE_TYPE_MAP, queueTypeVector);
+
```
will be changed

tomer-israel commented on tests/test_flex_counters.py:
```diff
@@ -110,3 +110,86 @@ def test_flex_counters(self, dvs, counter_type):
+        oid_list = []
+        counters_queue_map = self.counters_db.get_entry("COUNTERS_QUEUE_NAME_MAP","")
+        
+        for i in range(0,100):
+            if 'Ethernet0:%d'%i in counters_queue_map:
+                oid_list.append(counters_queue_map['Ethernet0:%d'%i])
+            else:
+                break
```
i changed it to while loop - break loop when queue id dont exist

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5418,6 +5498,41 @@ void PortsOrch::generatePriorityGroupMap()
+    m_counter_db->hdel(COUNTERS_PG_PORT_MAP, pgPortVector);
+    m_counter_db->hdel(COUNTERS_PG_INDEX_MAP, pgIndexVector);
+
+    CounterCheckOrch::getInstance().removePort(port);
```
I will leave it like that since the add function is doing the same (addPort).
refactor the generate function and the remove function will take too much effort for now

tomer-israel commented on tests/test_drop_counters.py:
```diff
@@ -654,6 +655,89 @@ def test_removeInvalidDropReason(self, dvs, testlog):
+        self.create_drop_counter(name2, PORT_EGRESS_DROPS)
+        self.add_drop_reason(name2, reason2)
+         
+        time.sleep(5)
```
added

stepanblyschak commented on orchagent/debugcounterorch.cpp:
```diff
@@ -34,13 +35,38 @@ DebugCounterOrch::DebugCounterOrch(DBConnector *db, const vector<string>& table_
+{
+    SWSS_LOG_ENTER();
+
+
```
reduntand empty line

stepanblyschak commented on tests/test_flex_counters.py:
```diff
@@ -110,3 +114,85 @@ def test_flex_counters(self, dvs, counter_type):
+        oid_list = []
+        counters_queue_map = self.counters_db.get_entry("COUNTERS_QUEUE_NAME_MAP", "")
+        
+        i = 0
+        while True:
+            if '%s:%d' % (PORT, i) in counters_queue_map:
+                oid_list.append(counters_queue_map['%s:%d' % (PORT, i)])
+                i += 1 
+            else:
+                break
```
This code looks too verbose and very C-style which makes it hard to understand without comments. Can we do this:

```python
oid_list = [oid for port_queue, oid in counters_queue_map.values() if port_queue.startswith(PORT)]
```

Which is self-explanatory: collect every oid in counters_queue_map if this oid belongs to PORT.

tomer-israel replied:
I changed it to simpler for loop.
one line loop as you suggested can be hard to understand and the for loop is simplier

stepanblyschak replied:
It is very common to see such list comprehension syntax in python. But this is IMO. Ok, if that is simpler for-loop

stepanblyschak commented on tests/test_drop_counters.py:
```diff
@@ -1,5 +1,6 @@
 import time
 import pytest
+import redis
```
Do we need redis still?

tomer-israel commented on orchagent/portsorch.cpp:
```diff
@@ -5557,7 +5672,8 @@ void PortsOrch::doTask(NotificationConsumer &consumer)
             {
-                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64, id);
+                /* consider changing it to notice */
+                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64 " - its ok to ignore this message if this port was just removed", id);
```
this will be removed

tomer-israel commented on tests/test_flex_counters.py:
```diff
@@ -110,3 +114,85 @@ def test_flex_counters(self, dvs, counter_type):
+                oid_list.append(counters_queue_map['%s:%d' % (PORT, i)])
+                i += 1 
+            else:
+                break
```
i changed it to simpler for loop

stepanblyschak commented on orchagent/portsorch.cpp:
```diff
@@ -5557,7 +5677,7 @@ void PortsOrch::doTask(NotificationConsumer &consumer)
             if (!getPort(id, port))
             {
-                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64, id);
+                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64 ", id);
```
"Failed to get port object for port id 0x%" PRIx64 "
The last " is redundant?

tomer-israel replied:
i revert it to original code.
this change will be pushed in a different PR since it is not related to counters changes

stepanblyschak commented on orchagent/portsorch.cpp:
```diff
@@ -5557,7 +5677,7 @@ void PortsOrch::doTask(NotificationConsumer &consumer)
             if (!getPort(id, port))
             {
-                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64, id);
+                SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64 ", id);
```
Now I am confused with this.
Originally, there was a comment saying it is actually a notice not an error. So I am wondering, if we catch this error in our regression tests - the test will fail?

tomer-israel replied:
after discussion with Dror we decided to add this change in a differetn PR since it's not really related to the counters changes.

stepanblyschak commented on tests/test_drop_counters.py:
```diff
@@ -156,30 +163,30 @@ def test_deviceCapabilitiesTablePopulated(self, dvs, testlog):
         """
         self.setup_db(dvs)
-
+  
```
Still see trailing spaces

stepanblyschak commented on orchagent/orchdaemon.cpp:
```diff
@@ -44,6 +44,7 @@ NatOrch *gNatOrch;
 MlagOrch *gMlagOrch;
 IsoGrpOrch *gIsoGrpOrch;
 MACsecOrch *gMacsecOrch;
+DebugCounterOrch *gDebugCounterOrch;
```
Is this addressed?

tomer-israel commented on tests/test_flex_counters.py:
```diff
@@ -110,3 +114,59 @@ def test_flex_counters(self, dvs, counter_type):
+        dvs.get_asic_db().wait_for_deleted_entry("ASIC_STATE:SAI_OBJECT_TYPE_PORT", port_oid)
+        
+        # verify counters were removed from flex counter table
+        for oid in oid_list:
```
you right it's empty, i forgot to add the oid to the oid_list in the previous for loop
i fixed it

smaheshm committed:
[CRM][MPLS] Fix the mpls nexthop CRM attribute (#2008)

pettershao-ragilenetworks committed:
Gcov for swss daemon (#1737)

* [gcov] swss support gcov
* Introduce gcov-related automation script to test coverage rate for swss module
Signed-off-by: pettershao-ragilenetworks <pettershao@ragilenetworks.com>

ayurkiv-nvda committed:
[switchorch] Implement VXLAN src port range feature  (#1959)

* Implement VXLAN src port range feature configuring
Signed-off-by: Andriy Yurkiv <ayurkiv@nvidia.com>

TACappleman committed:
[qos] Add EXP to TC map support (#1954)

* [Mpls/qos] Add EXP to TC map support

lguohan committed:
[ci]: use native arm64 and armhf pool (#2013)

Signed-off-by: Guohan Lu <lguohan@gmail.com>

theasianpianist committed:
[pytest]: Ignore errors deleting host ifs (#2005)

The delete command for host interfaces during virtual server set up can occasionally fail if the interface has been auto-deleted before the command is run. This causes cascading failures for the VS test suite. Ignore such an error if this occurs.

Signed-off-by: Lawrence Lee <lawlee@microsoft.com>

prsunny committed:
[Mux orch] set default as standby, change mux orch priority (#2010)

*Set default status for each mux port to standby state during init.
*Change mux orch priority to process entries before neighbor orch

tomer-israel commented:
close and create external PR
